### PR TITLE
Change linear op rewrite pattern 

### PR DESCRIPTION
**Pull request by ddilbazTT**
### Ticket
https://github.com/tenstorrent/tt-mlir/issues/5813
https://github.com/tenstorrent/tt-mlir/issues/6016

### Problem description
[During uplift 2 weeks ago](https://github.com/tenstorrent/tt-mlir/pull/5744), default tt-metal/ttnn linear op behaviour has changed. linear op is now mapping [PyTorch Linear](https://docs.pytorch.org/docs/stable/generated/torch.nn.Linear.html) This means that linear op bias needs to be 1d (or view, like 1x1x64)

@ctodTT relaxed the conditions to linear op workaround so that all linear ops would be broken down into matmul and add. However, this means no linear op is running as a linear op, even if they have 1d bias.

### What's changed
Linear op workaround is changed such that:
- only linear ops with non 1d bias or batched b input are broken down into matmul + add
- activation is passed from linear to matmul op correctly (previously, this was not passed)
- Tests marked unsupported or skipped are reinstated. 

### Checklist
- [Performance Benchmark](https://github.com/tenstorrent/tt-xla/actions/runs/19769666488 ): 
    - Errors seen are the same errors seen in other performance benchmark runs. I.e. (Circular buffer error, cv2 not found)
- [Model test push](https://github.com/tenstorrent/tt-xla/actions/runs/19769681564)
- [Basic Test](https://github.com/tenstorrent/tt-xla/actions/runs/19769684308)
- [Model Test Passing](https://github.com/tenstorrent/tt-xla/actions/runs/19772587791)

### CODE REVIEW

**Comment by ddilbazTT** [comment]
@sdjordjevicTT @jserbedzijaTT I believe this PR would address missing activations in linears with activations. However, I believe you might also consider changing the linear with activation fusing pattern to absorb the repeat that comes before. That way, they would remain as linear with 1d bias.

**Comment by ddilbazTT** [comment]
> Does this PR solve #5858?

Yes, I'll leave a comment there

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp [hunk 1]
```diff
@@ -91,11 +153,18 @@ LinearOpRewritePattern::matchAndRewrite(ttnn::LinearOp srcOp,
...
+    activationAttr = rewriter.getStringAttr(activation.value());
+  }
+
   MatmulOp matmulOp = rewriter.create<ttnn::MatmulOp>(
       ttmlir::utils::appendLocationSuffix(srcOp.getLoc(), "_decomp_matmul"),
       matmulOutputType, srcOp.getA(), srcOp.getB(), srcOp.getTransposeA(),
       srcOp.getTransposeB(),
-      /*matmul_program_config=*/mlir::Attribute(), /*activation=*/nullptr);
+      /*matmul_program_config=*/mlir::Attribute(),
+      /*activation=*/activationAttr);
```
Comment (mtopalovicTT) [inline lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: I'm not sure if this is correct. Activation is usually last op in the chain. For example we have matmul fallowed by add which is followed by some activation function.

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp [hunk 2]
```diff
@@ -14,20 +14,45 @@
...
+
+  for (int64_t dim : biasShape) {
+    if (dim > 1) {
+      nonUnitDims++;
+    }
+  }
+
+  // Bias is effectively 1D if it has at most one non-unit dimension
+  return nonUnitDims <= 1;
+}
```
Comment (mtopalovicTT) [inline lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]:
```suggestion
  return rank > 1 && llvm::any_of(biasShape.drop_back(1),
                                  [](int64_t dim) { return dim > 1; });
```

Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]: I don't think this is equivalent to the original code. I changed it to 
```
// Helper function to check if bias is effectively 1D
// Returns true if bias has only one non-unit dimension (e.g., <64>, <1x64>,
// <1x1x64>)
static bool isEffectively1DBias(TypedValue<RankedTensorType> bias) {
  if (!bias) {
    return false;
  }

  RankedTensorType biasType = bias.getType();
  if (!biasType) {
    return false;
  }

  auto biasShape = biasType.getShape();
  
  // Count non-unit dimensions.
  return llvm::count_if(biasShape, [](int64_t dim) { return dim > 1; }) <= 1;
}
```
instead.

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp (hunk 1 above)
Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: I'm passing linear activation attribute to matmul activation attribute when linear is decomposed. Is that wrong?

Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: @sdjordjevicTT could you also review this?

Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: Also @jserbedzijaTT - could you check if I'm passing in the activations correctly?

Comment (jserbedzijaTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: Looks like you are trying to rewrite activation(linear) into activation(matmul) + bias which is not mathematically correct, you should wrap both matmul and bias into the same activation function like activation(matmul + bias)

**Comment by jserbedzijaTT** [comment]
A couple of weeks ago, I created an issue on the metal side to change the Activation parameter of matmul from a string to UnaryWithParam, and it looks like they implemented it:

```cpp
using Activation = std::variant<std::string, UnaryWithParam>;

struct MatmulOperation {
    static Tensor invoke(
        const Tensor& input_tensor_a,
        const Tensor& input_tensor_b,
        bool transpose_a = false,
        bool transpose_b = false,
        const std::optional<const MemoryConfig>& memory_config = std::nullopt,
        std::optional<const DataType> dtype = std::nullopt,
        const std::optional<const MatmulProgramConfig>& program_config = std::nullopt,
        const std::optional<const Activation>& activation = std::nullopt,
        std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt,
        std::optional<const CoreGrid> core_grid = std::nullopt,
        const std::optional<const tt::tt_metal::Tile>& output_tile = std::nullopt,
        std::optional<Tensor> optional_output_tensor = std::nullopt,
        const std::optional<const GlobalCircularBuffer>& global_cb = std::nullopt,
        const std::optional<tt::tt_metal::SubDeviceId>& sub_device_id = std::nullopt);
};
```

I’m not sure if it’s too much work to update matmul in TTNN to use this, but it would be a nice change.

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp (hunk 2 above)
Comment (mtopalovicTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]: that won't work in case 1x64x1x1. You need to drop last dim.

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp (hunk 1 above)
Comment (mtopalovicTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: it should be either
matmul -> activation
or 
matmul -> add -> activation

**Comment by ddilbazTT** [comment]
> A couple of weeks ago, I created an issue on the metal side to change the Activation parameter of matmul from a string to UnaryWithParam, and it looks like they implemented it:
> 
> ```c++
> using Activation = std::variant<std::string, UnaryWithParam>;
> 
> struct MatmulOperation {
>     static Tensor invoke(
>         const Tensor& input_tensor_a,
>         const Tensor& input_tensor_b,
>         bool transpose_a = false,
>         bool transpose_b = false,
>         const std::optional<const MemoryConfig>& memory_config = std::nullopt,
>         std::optional<const DataType> dtype = std::nullopt,
>         const std::optional<const MatmulProgramConfig>& program_config = std::nullopt,
>         const std::optional<const Activation>& activation = std::nullopt,
>         std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt,
>         std::optional<const CoreGrid> core_grid = std::nullopt,
>         const std::optional<const tt::tt_metal::Tile>& output_tile = std::nullopt,
>         std::optional<Tensor> optional_output_tensor = std::nullopt,
>         const std::optional<const GlobalCircularBuffer>& global_cb = std::nullopt,
>         const std::optional<tt::tt_metal::SubDeviceId>& sub_device_id = std::nullopt);
> };
> ```
> 
> I’m not sure if it’s too much work to update matmul in TTNN to use this, but it would be a nice change.

Is it ok if it's a different PR? I think this should be about cleaning up ttnn linear op workaround.

**Comment by jserbedzijaTT** [comment]
> > A couple of weeks ago, I created an issue on the metal side to change the Activation parameter of matmul from a string to UnaryWithParam, and it looks like they implemented it:
> > ```c++
> > using Activation = std::variant<std::string, UnaryWithParam>;
> > 
> > struct MatmulOperation {
> >     static Tensor invoke(
> >         const Tensor& input_tensor_a,
> >         const Tensor& input_tensor_b,
> >         bool transpose_a = false,
> >         bool transpose_b = false,
> >         const std::optional<const MemoryConfig>& memory_config = std::nullopt,
> >         std::optional<const DataType> dtype = std::nullopt,
> >         const std::optional<const MatmulProgramConfig>& program_config = std::nullopt,
> >         const std::optional<const Activation>& activation = std::nullopt,
> >         std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt,
> >         std::optional<const CoreGrid> core_grid = std::nullopt,
> >         const std::optional<const tt::tt_metal::Tile>& output_tile = std::nullopt,
> >         std::optional<Tensor> optional_output_tensor = std::nullopt,
> >         const std::optional<const GlobalCircularBuffer>& global_cb = std::nullopt,
> >         const std::optional<tt::tt_metal::SubDeviceId>& sub_device_id = std::nullopt);
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I’m not sure if it’s too much work to update matmul in TTNN to use this, but it would be a nice change.
> 
> Is it ok if it's a different PR? I think this should be about cleaning up ttnn linear op workaround.

Yep no worries it just came to mind when I saw Activation attribute being passed to matmul :D

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp (hunk 1 above)
Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: I just changed it, I would appreciate a look. In terms of activations, we are talking about: `relu`, `sigmoid`, `gelu`, `tanh` right? Is there any that I'm missing? I just updated `lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp` 

```
  // Step 3: If original linear op had activation, apply it to the add result.
  if (srcOp.getActivation()) {
    std::string activationStr = srcOp.getActivation()->str();
    Value activationResult;

    if (activationStr == "relu") {
      activationResult = rewriter.create<ttnn::ReluOp>(
          ttmlir::utils::appendLocationSuffix(srcOp.getLoc(), "_decomp_relu"),
          outputType, addOp.getResult());
    } else if (activationStr == "sigmoid") {
      activationResult = rewriter.create<ttnn::SigmoidOp>(
          ttmlir::utils::appendLocationSuffix(srcOp.getLoc(),
                                              "_decomp_sigmoid"),
          outputType, addOp.getResult());
    } else if (activationStr == "gelu") {
      activationResult = rewriter.create<ttnn::GeluOp>(
          ttmlir::utils::appendLocationSuffix(srcOp.getLoc(), "_decomp_gelu"),
          outputType, addOp.getResult());
    } else if (activationStr == "tanh") {
      activationResult = rewriter.create<ttnn::TanhOp>(
          ttmlir::utils::appendLocationSuffix(srcOp.getLoc(), "_decomp_tanh"),
          outputType, addOp.getResult());
    } else {
      matmulOp.emitError()
          << "Unsupported activation type in LinearOp decomposition: "
          << activationStr;
      return failure();
    }
    rewriter.replaceOp(srcOp, activationResult);
  } else {
    rewriter.replaceOp(srcOp, addOp.getResult());
  }
```

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp (hunk 2 above)
Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]: Just to be clear - isn't `1x64x1x1` a valid bias shape?

Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]: Doesn't this check if there is a dim greater than 1 except the last dim? So it would return true for dim shape 32x32, for example.

Comment (mtopalovicTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]: Only last dim should be > 1, otherwise it will be broadcasted.

> Doesn't this check if there is a dim greater than 1 except the last dim? So it would return true for dim shape 32x32, for example.

My suggestion yes, but you can adjust it to your needs. I only wanted to show piece of code that checks if any dim except last one is > 1.

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp [hunk 3]
```diff
@@ -90,21 +145,54 @@ LinearOpRewritePattern::matchAndRewrite(ttnn::LinearOp srcOp,
...
-  rewriter.replaceOp(srcOp, addOp.getResult());
+  // Step 3: If original linear op had activation, apply it to the add result.
+  if (srcOp.getActivation()) {
+    std::string activationStr = srcOp.getActivation()->str();
+    Value activationResult;
+
+    if (activationStr == "relu") {
+      activationResult = rewriter.create<ttnn::ReluOp>(
+          ttmlir::utils::appendLocationSuffix(srcOp.getLoc(), "_decomp_relu"),
+          outputType, addOp.getResult());
```
Comment (mtopalovicTT) [inline lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: You can use this builder for all operations instead of doing them separately.

  /// Creates an operation with the given fields.
  Operation *create(Location loc, StringAttr opName, ValueRange operands,
                    TypeRange types = {},
                    ArrayRef<NamedAttribute> attributes = {},
                    BlockRange successors = {},
                    MutableArrayRef<std::unique_ptr<Region>> regions = {});

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp (hunk 2 above)
Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]: So I tried a few bias shapes:
- `<2x1x1>` - does not run because batch_size > 1
- `<1x34x1>` - runs on silicon
I think I need to add a check for batch_size > 1 when rank > 1 but otherwise, I think the function I have is correct. Let me know.

Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp:50]: I think you're right, my bad for making this convoluted. I should only check last dim is not 1 and everything else is 1. I think with your suggestion I need to invert the return value or change the name of the function to `notEffectively1DBias` or something. However, <1x34x1> is not really an accurate linear bias if we're referencing torch, even if it can run on tt silicon.

File: lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp (hunk 3 above)
Comment (ddilbazTT) [reply lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp]: Great suggestion, thanks!
